core/services/pg/locked_db.go

```diff
@@ -0,0 +1,129 @@
+package pg
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+
+	"github.com/smartcontractkit/chainlink/core/config"
+	"github.com/smartcontractkit/chainlink/core/logger"
+	"github.com/smartcontractkit/chainlink/core/static"
+	"github.com/smartcontractkit/sqlx"
+)
+
+// LockedDB bounds DB connection and DB locks.
+type LockedDB interface {
+	Open(ctx context.Context) error
+	Close() error
+	DB() *sqlx.DB
+}
+
+type lockedDb struct {
+	cfg          config.GeneralConfig
+	lggr         logger.Logger
+	db           *sqlx.DB
+	leaseLock    LeaseLock
+	advisoryLock AdvisoryLock
+}
+
+// NewLockedDB creates a new instance of LockedDB
+func NewLockedDB(cfg config.GeneralConfig, lggr logger.Logger) LockedDB {
+	return &lockedDb{
+		cfg:  cfg,
+		lggr: lggr.Named("LockedDB"),
+	}
+}
+
+// Open function connects to DB and acquires DB locks based on configuration.
+// If any of the steps fails or ctx is cancelled, it reverts everything.
+// This is a blocking function and it may execute long due to DB locks acquisition.
+// NOT THREAD SAFE
+func (l *lockedDb) Open(ctx context.Context) (err error) {
+	// If Open succeeded previously, db will not be nil
+	if l.db != nil {
+		return errors.New("calling Open() twice")
+	}
+
+	// Step 1: open DB connection
+	l.db, err = openDB(l.cfg, l.lggr)
+	if err != nil {
+		// l.db will be nil in case of error
+		return errors.Wrap(err, "failed to open db")
+	}
+	revert := func() {
+		// Let Open() return the actual error, while l.Close() error is just logged.
+		if err2 := l.Close(); err2 != nil {
+			l.lggr.Errorf("failed to cleanup LockedDB: %v", err2)
+		}
+	}
+
+	// Step 2: acquire DB locks
+	lockingMode := l.cfg.DatabaseLockingMode()
+	l.lggr.Debugf("Using database locking mode: %s", lockingMode)
+
+	// Take the lease before any other DB operations
+	switch lockingMode {
+	case "lease", "dual":
+		l.leaseLock = NewLeaseLock(l.db, l.cfg.AppID(), l.lggr, l.cfg.LeaseLockRefreshInterval(), l.cfg.LeaseLockDuration())
+		if err = l.leaseLock.TakeAndHold(ctx); err != nil {
+			defer revert()
+			return errors.Wrap(err, "failed to take initial lease on database")
+		}
+	}
+
+	// Try to acquire an advisory lock to prevent multiple nodes starting at the same time
+	switch lockingMode {
+	case "advisorylock", "dual":
+		l.advisoryLock = NewAdvisoryLock(l.db, l.cfg.AdvisoryLockID(), l.lggr, l.cfg.AdvisoryLockCheckInterval())
+		if err = l.advisoryLock.TakeAndHold(ctx); err != nil {
+			defer revert()
```

Comment:
revert should include undoing the lease lock

Reply:
`revert()` calls `LockDB.Close()` which in turn releases all locks and closes DB.